File: javascript/json-transform/src/JsonTransformer.ts
import { Transformer } from "./Transformer";
import { JsonTransformerFunction } from "./JsonTransformerFunction";
import { ParameterResolver } from "./ParameterResolver";
import { createPayloadResolver, isNullOrUndefined, toObjectFieldPath } from "./JsonHelpers";
import transformerFunctions, { TransformerFunctionsAdapter } from "./TransformerFunctions";
import JsonElementStreamer from "./JsonElementStreamer";
import BigNumber from "bignumber.js";
import { BigDecimal } from "./functions/common/FunctionHelpers";

export class JsonTransformer implements Transformer {
  static readonly OBJ_DESTRUCT_KEY = "*";
  static readonly NULL_VALUE = "#null";

  private transformerFunctions: TransformerFunctionsAdapter;
  private definition: any;
  private JSON_TRANSFORMER: JsonTransformerFunction;

  public constructor(definition: any, functionsAdapter?: TransformerFunctionsAdapter) {
    this.transformerFunctions = functionsAdapter ?? transformerFunctions;
    this.definition = definition;
    this.JSON_TRANSFORMER = {
      transform: this.fromJsonElement.bind(this),
    };
  }

  /**
   * Transforms the payload using the transformer definition
   *
   * @param payload               The payload to transform
   * @param additionalContext     (optional) Additional context to use in the transformation
   * @returns A promise to the transformed payload
   */
  async transform(payload: any = null, additionalContext: Record<string, any> = {}) {
    if (isNullOrUndefined(this.definition)) {
      return null;
    }
    const resolver: ParameterResolver = createPayloadResolver(payload, additionalContext);
    return this.fromJsonElement("$", this.definition, resolver, false);
  }

  protected async fromJsonPrimitive(
    path: string,
    definition: any,
    resolver: ParameterResolver,
    allowReturningStreams: boolean,
  ): Promise<any> {
    if (typeof definition !== "string") {
      return definition ?? null;
    }
    try {
      // test for inline function (e.g. $$function:...)
      const match = await this.transformerFunctions.matchInline(path, definition, resolver, this.JSON_TRANSFORMER);
      if (match != null) {
        const matchResult = match.getResult();
        if (matchResult instanceof JsonElementStreamer) {
          return allowReturningStreams ? matchResult : await matchResult.toJsonArray();
        }
        return match.getResult();
      }
      // jsonpath / context
      return resolver.get(definition);
    } catch (ignored: any) {
      // console.warn(`Failed getting value for ${definition}`, ignored);
      return null;
    }
  }

  protected async fromJsonObject(
    path: string,
    definition: any,
    resolver: ParameterResolver,
    allowReturningStreams: boolean,
  ): Promise<any> {
    const match = await this.transformerFunctions.matchObject(path, definition, resolver, this.JSON_TRANSFORMER);
    if (match != null) {
      const res = match.getResult();
      if (res instanceof JsonElementStreamer) {
        return allowReturningStreams ? res : await res.toJsonArray();
      }
      return match.getResult();
    }

    let result: Record<string, any> = {};
    var hasObjectDestruction = Object.prototype.hasOwnProperty.call(definition, JsonTransformer.OBJ_DESTRUCT_KEY);
    if (hasObjectDestruction) {
      const val = definition[JsonTransformer.OBJ_DESTRUCT_KEY];
      const res = await this.fromJsonElement(path + '["*"]', val, resolver, false);
      if (res != null) {
        const isArray = Array.isArray(val);
        if (isArray && Array.isArray(res)) {
          for (let i = 0; i < res.length; i++) {
            if (typeof res[i] === "object") {
              Object.assign(result, res[i]);
            }
          }
        } else if (typeof res === "object") {
          result = res;
        } else {
          result[JsonTransformer.OBJ_DESTRUCT_KEY] = res;
        }
      }
    }

    for (const key in definition) {
      if (key === JsonTransformer.OBJ_DESTRUCT_KEY) continue;
      const localValue = definition[key];
      if (localValue === JsonTransformer.NULL_VALUE) {
        if (hasObjectDestruction) {
          // don't define key if #null was used
          // might already exist, so try removing it
          delete result[key];
        } else {
          result[key] = null;
        }
        continue;
      }
      const value = await this.fromJsonElement(path + toObjectFieldPath(key), localValue, resolver, false);
      if (
        !isNullOrUndefined(value) ||
        Object.prototype.hasOwnProperty.call(result, key) /* we allow overriding with null*/
      ) {
        result[key] = value;
      }
    }

    return result;
  }

  protected async fromJsonElement(
    path: string,
    definition: any,
    resolver: ParameterResolver,
    allowReturningStreams: boolean,
  ): Promise<any> {
    if (isNullOrUndefined(definition)) {
      return null;
    }
    if (Array.isArray(definition)) {
      return Promise.all(definition.map((d: any, i) => this.fromJsonElement(`${path}[${i}]`, d, resolver, false)));
    }
    if (typeof definition === "object" && !(definition instanceof BigNumber || definition instanceof BigDecimal)) {
      return this.fromJsonObject(path, definition, resolver, allowReturningStreams);
    }
    return this.fromJsonPrimitive(path, definition, resolver, allowReturningStreams);
  }

  /**
   * Gets the transformer definition
   */
  getDefinition() {
    return this.definition;
  }
}

File: java/json-transform/src/main/java/co/nlighten/jsontransform/JsonTransformer.java
package co.nlighten.jsontransform;

import co.nlighten.jsontransform.adapters.JsonAdapter;

import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * A transformer is used to transform data from one layout to another
 */
public class JsonTransformer implements Transformer {

    static final String OBJ_DESTRUCT_KEY = "*";
    static final String NULL_VALUE = "#null";

    private final JsonAdapter<?, ?, ?> adapter;
    protected final Object definition;

    private static final TransformerFunctionsAdapter DEFAULT_TRANSFORMER_FUNCTIONS = new TransformerFunctions();
    private final JsonTransformerFunction JSON_TRANSFORMER;
    private final TransformerFunctionsAdapter transformerFunctions;

    /**
     * Creates a new JSON transformer from definition
     *
     * @param definition       The transformer definition
     * @param adapter          (optional) A specific JSON implementation adapter (otherwise uses the configured default)
     * @param functionsAdapter (optional) A specific transformer functions adapter (otherwise uses the default)
     */
    public JsonTransformer(
            final Object definition,
            final JsonAdapter<?, ?, ?> adapter,
            final TransformerFunctionsAdapter functionsAdapter) {
        this.adapter = adapter != null ? adapter : JsonTransformerConfiguration.get().getAdapter();
        this.transformerFunctions = functionsAdapter != null ? functionsAdapter : DEFAULT_TRANSFORMER_FUNCTIONS;
        this.definition = this.adapter.wrap(definition);
        this.JSON_TRANSFORMER = this::fromJsonElement;
    }

    /**
     * Creates a new JSON transformer from definition
     *
     * @param definition       The transformer definition
     * @param adapter          (optional) A specific JSON implementation adapter (otherwise uses the configured default)
     */
    public JsonTransformer(
            final Object definition,
            final JsonAdapter<?, ?, ?> adapter) {
        this(definition, adapter, null);
    }

    /**
     * Creates a new JSON transformer from definition
     *
     * @param definition       The transformer definition
     * @param functionsAdapter (optional) A specific transformer functions adapter (otherwise uses the default)
     */
    public JsonTransformer(
            final Object definition,
            final TransformerFunctionsAdapter functionsAdapter) {
        this(definition, null, functionsAdapter);
    }

    /**
     * Creates a new JSON transformer from definition
     *
     * @param definition       The transformer definition
     */
    public JsonTransformer(final Object definition) {
        this(definition, DEFAULT_TRANSFORMER_FUNCTIONS);
    }

    /**
     * Transforms the payload using the transformer definition
     *
     * @param payload               The payload to transform
     * @param additionalContext     (optional) Additional context to use in the transformation
     * @param unwrap                (optional) Unwrap the result to POJO from the used JSON implementation (default is false)
     */
    @Override
    public Object transform(Object payload, Map<String, Object> additionalContext, boolean unwrap) {
        if (definition == null) {
            return unwrap ? null : adapter.jsonNull();
        }
        var resolver = adapter.createPayloadResolver(payload, additionalContext, false);
        var result = fromJsonElement("$", definition, resolver, false);
        return unwrap ? adapter.unwrap(result) : result;
    }


    protected Object fromJsonPrimitive(String path, Object definition, ParameterResolver resolver, boolean allowReturningStreams) {
        if (!adapter.isJsonString(definition))
            return definition;
        try {
            var val = adapter.getAsString(definition);
            // test for inline function (e.g. $$function:...)
            var match = transformerFunctions.matchInline(adapter, path, val, resolver, JSON_TRANSFORMER);
            if (match != null) {
                var matchResult = match.result();
                if (matchResult instanceof JsonElementStreamer streamer) {
                    return allowReturningStreams ? streamer : streamer.toJsonArray();
                }
                return adapter.wrap(matchResult);
            }
            // jsonpath / context
            var value = resolver.get(val);
            return adapter.wrap(value);
        } catch (Exception ee) {
            //logger.trace("Failed getting value for " + definition, ee);
            return adapter.jsonNull();
        }
    }

    protected Object fromJsonObject(String path, Object definition, ParameterResolver resolver, boolean allowReturningStreams) {
        var match = transformerFunctions.matchObject(adapter, path, definition, resolver, JSON_TRANSFORMER);
        if (match != null) {
            var res = match.result();
            return res instanceof JsonElementStreamer s
                    ? (allowReturningStreams ? s : s.toJsonArray())
                    : adapter.wrap(res);
        }

        var result = adapter.createObject();
        var hasObjectDestruction = adapter.has(definition, OBJ_DESTRUCT_KEY);
        if (hasObjectDestruction) {
            var val = adapter.get(definition, OBJ_DESTRUCT_KEY);
            var res = fromJsonElement(path + "[\"*\"]", val, resolver, false);
            if (res != null) {
                var isArray = adapter.isJsonArray(val);
                if (isArray && adapter.isJsonArray(res)) {
                    for (var x: adapter.asIterable(res)) {
                        if (adapter.isJsonObject(x)) {
                            for (Map.Entry<String, ?> e : adapter.entrySet(x)) {
                                adapter.add(result, e.getKey(), e.getValue());
                            }
                        }
                    }
                } else if (adapter.isJsonObject(res)) {
                    // override the base object with resolved one
                    result = res;
                } else {
                    adapter.add(result, OBJ_DESTRUCT_KEY, res);
                }
            }
        }

        for (Map.Entry<String, ?> kv : adapter.entrySet(definition)) {
            if (kv.getKey().equals(OBJ_DESTRUCT_KEY)) continue;
            var localKey = kv.getKey();
            var localValue = kv.getValue();
            if (adapter.isJsonString(localValue) && adapter.getAsString(localValue).equals(NULL_VALUE)) {
                if (hasObjectDestruction) {
                    // don't define key if #null was used
                    // might already exist, so try removing it
                    adapter.remove(result, localKey);
                } else {
                    adapter.add(result, localKey, adapter.jsonNull());
                }
                continue;
            }
            var value = fromJsonElement(
                    path + JsonTransformerUtils.toObjectFieldPath(adapter, localKey),
                    localValue, resolver, false);
            if (!adapter.isNull(value) || adapter.has(result, localKey) /* we allow overriding with null*/) {
                adapter.add(result, localKey, value);
            }
        }

        return result;
    }

    protected Object fromJsonElement(String path, Object definition, ParameterResolver resolver, boolean allowReturningStreams) {
        if (adapter.isNull(definition))
            return adapter.jsonNull();
        if (adapter.isJsonArray(definition)) {
            var result = adapter.createArray();
            var index = new AtomicInteger(0);
            adapter.stream(definition, false)
                    .map(d -> fromJsonElement(path + "[" + index.getAndIncrement() + "]", d, resolver, false))
                    .forEachOrdered(item -> adapter.add(result, item));
            return result;
        }
        if (adapter.isJsonObject(definition)) {
            return fromJsonObject(path, definition, resolver, allowReturningStreams);
        }
        return fromJsonPrimitive(path, definition, resolver, allowReturningStreams);
    }

    /**
     * Gets the transformer definition
     */
    public Object getDefinition() {
        return definition;
    }
}
